pulp_ansible/app/tasks/collectionversion_index.py

```diff
@@ -0,0 +1,201 @@
+import logging
+from semantic_version import Version
+
+from django.db.models import Q
+
+from pulp_ansible.app.models import (
+    AnsibleDistribution,
+    AnsibleCollectionDeprecated,
+    AnsibleNamespace,
+    AnsibleRepository,
+    CollectionVersion,
+    CollectionVersionSignature,
+    CrossRepositoryCollectionVersionIndex as CVIndex,
+)
+
+
+log = logging.getLogger(__name__)
+
+
+def has_distribution(repository, repository_version):
+    """Is -any- distribution attached to this repo or repo version?"""
+    return AnsibleDistribution.objects.filter(
+        Q(repository=repository) | Q(repository_version=repository_version)
+    ).exists()
+
+
+def get_highest_version_string_from_cv_objects(cv_objects):
+    """Return the highest version as a string (preferring stable releases)."""
+    versions_strings = [x.version for x in cv_objects]
+    versions = [Version(x) for x in versions_strings]
+    versions = sorted(versions, reverse=True)
+    stable_versions = [x for x in versions if not x.prerelease]
+    if stable_versions:
+        return str(stable_versions[0])
+    return str(versions[0])
+
+
+def update_index(distribution=None, repository=None, repository_version=None, is_latest=False):
+    """Rebuild index by distribtion|repository|repositoryversion."""
+
+    # if the distro points at a specific repo version, we should use that in the index
+    # otherwise the index value for repository version should be null
+    use_repository_version = False
```

Comment:
I am not sure what is going on with all the commented out code below, but I am beginning to believe we should always specify a repository version when creating an index.

Reply:
The commented out code was mostly oriented around the mistaken belief that repository versions would always be "latest". However, the expectations have changed after yours and @newswangerd's reviews so the code was wrong. The additional tests clarified and assert the current behavior.

We can't always set a repository version because we need to ensure that (repository, no version, cv1) and (repository, repository version 1, cv1) are two unique results that come back from the viewset. I think it would have been more obvious if we had kept the distribution field in the index model, but we currently have 10's of thousands of "synclist" distributions pointing to the same repository in CRC so we had to get rid of that column to prevent possibly millions of duplicate entries.